libtenzir/builtins/operators/chart.cpp

```diff
@@ -133,87 +178,198 @@ class chart_operator final : public crtp_operator<chart_operator> {
+    for (const auto& item : cfg_) {
+      if (item.req == requirement::none) {
+        continue;
+      }
+      auto field_name = resolve_attribute_value(item, record_schema,
+                                                slice.schema().name(), ctrl);
+      if (not field_name) {
+        return false;
+      }
+      if (not verify_single_field(item, slice, *struct_array, record_schema,
+                                  *field_name, previous_values, ctrl)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  auto make_attributes(const table_slice& slice,
+                       previous_values_type& previous_values,
+                       operator_control_plane& ctrl) const
+    -> std::optional<std::vector<type::attribute_view>> {
+    const auto* record_schema_ptr = caf::get_if<record_type>(&slice.schema());
     if (not record_schema_ptr) {
       diagnostic::error("chart operator expects input to be a record")
         .emit(ctrl.diagnostics());
-      return {};
+      return std::nullopt;
     }
     const auto& record_schema = *record_schema_ptr;
-    std::vector<type::attribute_view> result;
-    // Go through `cfg_`, and append a corresponding attribute in `result`
-    for (const auto& attr : cfg_) {
-      auto visitor = detail::overload{
-        // `nth_field`:
-        // Value of the attribute is the name of the `f.index`th field.
-        // Default for --x-axis and --y-axis
-        [&](nth_field f) -> std::optional<std::string_view> {
-          if (record_schema.num_fields() <= f.index) {
-            diagnostic::warning("field at index {} not found in input "
-                                "(schema `{}`), but the chart operator "
-                                "expected it "
-                                "for `{}`",
-                                f.index, schema.name(), attr.first)
-              .note("from `{}`", name())
+    auto struct_array = to_record_batch(slice)->ToStructArray().ValueOrDie();
+    std::vector<type::attribute_view> result{};
+    for (const auto& item : cfg_) {
+      auto field_name = resolve_attribute_value(item, record_schema,
+                                                slice.schema().name(), ctrl);
+      if (not field_name) {
+        return std::nullopt;
+      }
+      if (not verify_single_field(item, slice, *struct_array, record_schema,
+                                  *field_name, previous_values, ctrl)) {
+        return std::nullopt;
+      }
+      result.emplace_back(item.key, *field_name);
+    }
+    return result;
+  }
+
+  auto
+  verify_single_field(const configuration_item& item, const table_slice& slice,
+                      const arrow::StructArray& struct_array,
+                      const record_type& record_schema,
+                      std::string_view field_name,
+                      previous_values_type& previous_values,
+                      operator_control_plane& ctrl) const -> bool {
```

Comment:
For checking the requirements in this function, should we check the requirements per schema id, or over all the data globally, i.e., do we care about ordering with every schema or ordering in the entire pipeline?

Reply:
I don't know. Can the frontend handle wildly different schemas when charting? If it can, per-schema checking makes sense.

Reply:
Right now, all results rendering in the frontend is per-schema. We eventually want to add a view that's across schemas, but that doesn't exist yet.

I think it's fine if the check is across schemas. It's slightly more strict than it needs to be right now, but there's no harm in having it, I think.